File: sportz-fastapi/src/schemas/commentary.py
"""
Pydantic validation schemas for commentary.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

# Query params
class ListCommentaryQuery(BaseModel):
    limit: Optional[int] = Field(default=None, ge=1, le=100)

# Request body
class CreateCommentaryBody(BaseModel):
    minute: Optional[int] = Field(default=None, ge=0)
    sequence: Optional[int] = None
    period: Optional[str] = None
    event_type: Optional[str] = Field(default=None, alias="eventType")
    actor: Optional[str] = None
    team: Optional[str] = None
    message: str = Field(min_length=1)
    metadata: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None

    model_config = {"populate_by_name": True}

# Response body
class CommentaryOut(BaseModel):
    id: int
    match_id: int = Field(serialization_alias="matchId")
    minute: Optional[int] = None
    sequence: Optional[int] = None
    period: Optional[str] = None
    event_type: Optional[str] = Field(default=None, serialization_alias="eventType")
    actor: Optional[str] = None
    team: Optional[str] = None
    message: str
    metadata: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    created_at: Optional[datetime] = Field(default=None, serialization_alias="createdAt")

    model_config = {"from_attributes": True, "populate_by_name": True}

File: sportz-gorilla/internal/validation/commentary.go
package validation

import (
	"errors"
	"fmt"
)

// CreateCommentaryBody mirrors createCommentarySchema (Zod) / CreateCommentaryBody (Pydantic).
type CreateCommentaryBody struct {
	Minute    *int           `json:"minute"`
	Sequence  *int           `json:"sequence"`
	Period    *string        `json:"period"`
	EventType *string        `json:"eventType"`
	Actor     *string        `json:"actor"`
	Team      *string        `json:"team"`
	Message   string         `json:"message"`
	Metadata  map[string]any `json:"metadata"`
	Tags      []string       `json:"tags"`
}

// Validate checks required fields and constraints.
func (b *CreateCommentaryBody) Validate() error {
	if b.Message == "" {
		return errors.New("message must not be empty")
	}
	if b.Minute != nil && *b.Minute < 0 {
		return fmt.Errorf("minute must be >= 0")
	}
	return nil
}

// ListCommentaryQuery mirrors listCommentaryQuerySchema.
type ListCommentaryQuery struct {
	Limit int
}

// ParseListCommentaryQuery parses and validates the `limit` query param.
func ParseListCommentaryQuery(rawLimit string) (ListCommentaryQuery, error) {
	limit := 10
	if rawLimit != "" {
		n := 0
		if _, err := fmt.Sscanf(rawLimit, "%d", &n); err != nil || n < 1 || n > 100 {
			return ListCommentaryQuery{}, errors.New("limit must be an integer between 1 and 100")
		}
		limit = n
	}
	return ListCommentaryQuery{Limit: limit}, nil
}